feat(docs): Add documentation for using records and typedef (#6382)
dart-lang/site-www: src/content/language/records.md

  ButtonItem({required this.label, required this.icon, this.onPressed});
  bool get hasOnpressed => onPressed != null;
}
```

Or to an [extension type][]:

```dart
extension type ButtonItem._(({String label, Icon icon, void Function()? onPressed}) _) {
  String get label => _.label;
  Icon get icon => _.icon;
  void Function()? get onPressed => _.onPressed;
  ButtonItem({required String label, required Icon icon, void Function()? onPressed})
      : this._((label: label, icon: icon, onPressed: onPressed));
  bool get hasOnpressed => _.onPressed != null;
}
```

And then create the list of button definitions using that type's constructors:

```dart
final List<ButtonItem> buttons =  [
  ButtonItem(
    label: "Button I",
    icon: const Icon(Icons.upload_file),
    onPressed: () => print("Action -> Button I"),
  ),
  ButtonItem(
    label: "Button II",
    icon: const Icon(Icons.info),
    onPressed: () => print("Action -> Button II"),
  )
];
```

Again, all while not needing to change the code that uses that list.

Changing any type does require the code using it to be very careful about
not making assumptions. A type alias does not offer any protection or guarantee

> for the code using it that the value is a record.

Reviewer: This sentence is a little hard to read. 
I'd maybe put a comma after "it".

> not making assumptions. A type alias does not offer any protection or guarantee
> for the code using it that the value is a record.

Reviewer 2: ```suggestion
not making assumptions. A type alias does not offer any protection or guarantee,
for the code using it as a reference, that the value being aliased is a record.
```

Extension types, also, offer little protection.

> Only a class can provide full abstraction and control over access to its state.

Reviewer: I wonder if "encapsulation" is a better word than "abstraction" here. Or maybe both, with "encapsulation" being a shorter way to say "control over access to its state".

> Only a class can provide full abstraction and control over access to its state.

Reviewer 2: ```suggestion
Only a class can provide full abstraction and encapsulation.
```

I like both!
